### PR TITLE
Avoid mutable default arguments

### PR DESCRIPTION
**Pull request by kmaehashi**
Fix arguments not fixed in #4419.

### CODE REVIEW

**Comment by niboshi** [comment]
How about using different default values from `None`?

**Comment by niboshi** [comment]
@kmaehashi I think it's better to use some singleton value (like `numpy._NoValue`).

**Comment by okuta** [comment]
How about using `()`?

**Comment by niboshi** [comment]
I don't think `()` is good because one may think it has some meaning.

**Comment by beam2d** [comment]
I prefer `'default'` in this case. It's much easier to use. We don't need to distinguish if the user passed the argument explicitly (distinguishing them is the purpose of `numpy._NoValue`).

**Comment by kmaehashi** [comment]
Currently `variable_style` of `build_computational_graph` behaves as follows:

* When dict is specified: use the dict as option
* When None is specified: use Chainer v1-style option (ie `{}`)
* When no value is specified: use default option (ie `{'shape': 'octagon', 'fillcolor': '#E0E0E0', 'style': 'filled'}`)

My proposal is:

* When dict is specified: use the dict as option
* When None is specified: use default option (`{'shape': 'octagon', 'fillcolor': '#E0E0E0', 'style': 'filled'}`)
(ie, users currently specifying Chaienr-v1 style (None) should specify `{}` instead.)

@niboshi -san's proposal is to prepare singleton immutable object to keep the current spec while removing the mutable default arguments.

**Comment by beam2d** [comment]
My suggestion was:
- When dict is specified: use the dict as option.
- When `None` is specified: use Chainer v1-style option, or raise an error (if we choose to remove this feature).
- When `'default'` is specified: use the default option.
- When nothing is specified: use the default option.
What I meant is that we do not need to distinguish between the last two cases.

**Comment by beam2d** [comment]
I rather thought a "hidden default value" is far more complex for users. For example, consider a user who wants to write a wrapper of this function that passes through this argument. If the function is using hidden default value semantics, the user has to emulate this behavior using again a hidden default value (possibly using the "hidden" value by accessing an private API or inventing another default value which needs to be translated in the wrapper function).

**Comment by kmaehashi** [comment]
Thanks for the discussion.
@niboshi If your point is to keep current spec as-is, I'd like to go with `'default'`. Do you think it's OK?

**Comment by niboshi** [comment]
Sorry for delay.
I think it's ok to proceed with `'default'`.

**Comment by kmaehashi** [comment]
Updated according to the discussion above.

File: chainer/computational_graph.py [hunk 1]
```diff
@@ -53,7 +55,9 @@ def label(self):
 
 class ComputationalGraph(object):
 
-    """Class that represents computational graph.
+    """__init__(self, nodes, edges, variable_style=_var_style, function_style=_func_style, rankdir='TB', remove_variable=False, show_name=True)
```
Comment (toslunar) [inline chainer/computational_graph.py]: Could you avoid `NOQA` as #6184?

Comment (kmaehashi) [reply chainer/computational_graph.py]: Thanks, fixed.

File: chainer/computational_graph.py [hunk 2]
```diff
@@ -190,10 +208,12 @@ def _skip_variable(nodes, edges):
 
 
 def build_computational_graph(
-        outputs, remove_split=True, variable_style=_var_style,
-        function_style=_func_style, rankdir='TB', remove_variable=False,
+        outputs, remove_split=True, variable_style=_default,
+        function_style=_default, rankdir='TB', remove_variable=False,
         show_name=True):
-    """Builds a graph of functions and variables backward-reachable from outputs.
+    """build_computational_graph(outputs, remove_split=True, variable_style=_var_style, function_style=_func_style, rankdir='TB', remove_variable=False, show_name=True)
```
Comment (toslunar) [inline chainer/computational_graph.py]: ditto 🙇

Comment (kmaehashi) [reply chainer/computational_graph.py]: fixed :bow:

File: chainer/computational_graph.py [hunk 3]
```diff
@@ -190,10 +208,14 @@ def _skip_variable(nodes, edges):
 
 
 def build_computational_graph(
-        outputs, remove_split=True, variable_style=_var_style,
-        function_style=_func_style, rankdir='TB', remove_variable=False,
+        outputs, remove_split=True, variable_style=_default,
+        function_style=_default, rankdir='TB', remove_variable=False,
```
Comment (niboshi) [inline chainer/computational_graph.py]: Let's use the explicit value `'default'` here and make it documented.

File: chainer/training/extensions/parameter_statistics.py [hunk 4]
```diff
@@ -8,8 +8,24 @@
...
+    'max': lambda x: backend.get_array_module(x).max(x),
+    'zeros': lambda x: backend.get_array_module(x).count_nonzero(x == 0),
+    'percentile': lambda x: backend.get_array_module(x).percentile(
+        x, (0.13, 2.28, 15.87, 50, 84.13, 97.72, 99.87))
+}
+
+
 class ParameterStatistics(extension.Extension):
-    """Trainer extension to report parameter statistics.
+    """__init__(self, links, statistics=_default_statistics, \
```
Comment (niboshi) [inline chainer/training/extensions/parameter_statistics.py]: Ditto.

File: chainer/computational_graph.py [hunk 5]
```diff
@@ -53,7 +55,11 @@ def label(self):
 
 class ComputationalGraph(object):
 
-    """Class that represents computational graph.
+    """__init__(self, nodes, edges, variable_style=_var_style, \
```
Comment (niboshi) [inline chainer/computational_graph.py]: Ditto.

File: chainer/computational_graph.py (hunk 3 above)
Comment (toslunar) [reply chainer/computational_graph.py]: Then I'd like to have the doc like
```
        function_style(dict or 'default'): Dot node style for function.
            The ``'default'`` style is
            ``{'shape': 'record', 'fillcolor': '#6495ED', 'style': 'filled'}``.
```

I'm wondering if we can do it without repeating the default values.

**Comment by kmaehashi** [comment]
@niboshi I'd like to discuss *API expansion* (`'default'` support) in another issue if needed. It's out of the scope of this PR. I'd also like to backport this.

**Comment by niboshi** [comment]
Hmm, in my understanding the purpose of using `str` argument is to make it explicit to users.

https://github.com/chainer/chainer/pull/4822#issuecomment-400913963
>I rather thought a "hidden default value" is far more complex for users. For example, consider a user who wants to write a wrapper of this function that passes through this argument. If the function is using hidden default value semantics, the user has to emulate this behavior using again a hidden default value (possibly using the "hidden" value by accessing an private API or inventing another default value which needs to be translated in the wrapper function).

Also I don't necessarily think this PR needs backporting, because it does not itself improve anything, but it's a safeguard for future bug introduction.

**Comment by toslunar** [comment]
If #4413 will not be backported, it doesn't look too bad to change v7 only.  Almost all the new PRs are into `master` branch.

**Comment by kmaehashi** [comment]
> Hmm, in my understanding the purpose of using str argument is to make it explicit to users.

Ok, then how about:
* renaming `_default` to `DEFAULT` so that users can use it
* make it documented
* do not backport
Or do you prefer to use `'default'` instead of `DEFAULT`?

> If #4413 will not be backported, it doesn't look too bad to change v7 only. Almost all the new PRs are into master branch.

I agree with this.

**Comment by niboshi** [comment]
@kmaehashi 
Do you mean introducing `chainer.DEFAULT`?
Hmm, I don't think it's pythonic (or Chainer-ish. I think `str` is more common for this kind of arguments).

**Comment by kmaehashi** [comment]
Updated according to the offline discussion.

File: chainer/training/extensions/parameter_statistics.py [hunk 6]
```diff
@@ -30,9 +41,12 @@ class ParameterStatistics(extension.Extension):
         links (~chainer.Link or iterable of ~chainer.Link): Link(s) containing
             the parameters to observe. The link is expected to have a ``name``
             attribute which is used as a part of the report key.
-        statistics (dict): Dictionary with function name to function mappings.
+        statistics (dict or 'default'): Dictionary with function name to
+            function mappings.
             The name is a string and is used as a part of the report key. The
             function is responsible for generating the statistics.
+            If the special value ``'default'`` is specified, use the default
+            statistics functions.
```
Comment (niboshi) [inline chainer/training/extensions/parameter_statistics.py]: Tiny suggestion:
```diff
- use the default statistics functions.
+ the default statistics functions will be used.
```

File: chainer/computational_graph.py [hunk 7]
```diff
@@ -205,9 +230,14 @@ def build_computational_graph(
             :class:`~chainer.FunctionNode` object.
         remove_split(bool): It must be ``True``. This argument is left for
             backward compatibility.
-        variable_style(dict): Dot node style for variable.
-            Possible keys are 'shape', 'color', 'fillcolor', 'style', and etc.
-        function_style(dict): Dot node style for function.
+        variable_style(dict or 'default'): Dot node style for variable.
+            Possible keys are 'shape', 'color', 'fillcolor', 'style' etc.
+            If the special value ``'default'`` is specified, use the default
```
Comment (niboshi) [inline chainer/computational_graph.py]: ditt.

File: chainer/computational_graph.py [hunk 8]
```diff
@@ -205,9 +230,14 @@ def build_computational_graph(
...
+            configuration.
+        function_style(dict or 'default'): Dot node style for function.
+            Possible keys are 'shape', 'color', 'fillcolor', 'style' etc.
+            If the special value ``'default'`` is specified, use the default
```
Comment (niboshi) [inline chainer/computational_graph.py]: ditto.

File: chainer/training/extensions/parameter_statistics.py [hunk 9]
```diff
@@ -52,17 +79,9 @@ class ParameterStatistics(extension.Extension):
...
-        'std': lambda x: backend.get_array_module(x).std(x),
-        'min': lambda x: backend.get_array_module(x).min(x),
-        'max': lambda x: backend.get_array_module(x).max(x),
-        'zeros': lambda x: backend.get_array_module(x).count_nonzero(x == 0),
-        'percentile': lambda x: backend.get_array_module(x).percentile(
-            x, (0.13, 2.28, 15.87, 50, 84.13, 97.72, 99.87))
-    }
-
-    def __init__(self, links, statistics=default_statistics,
+    default_statistics = _default_statistics
```
Comment (niboshi) [inline chainer/training/extensions/parameter_statistics.py]: Is this attribute needed?

File: chainer/training/extensions/parameter_statistics.py (hunk 6 above)
Comment (kmaehashi) [reply chainer/training/extensions/parameter_statistics.py]: Fixed as suggested.

File: chainer/training/extensions/parameter_statistics.py (hunk 9 above)
Comment (kmaehashi) [reply chainer/training/extensions/parameter_statistics.py]: I think its convenient for users.
Let's discuss the topic in the separate issue if needed.

Comment (niboshi) [reply chainer/training/extensions/parameter_statistics.py]: OK.

**Comment by niboshi** [comment]
Jenkins, test this please

**Pull request merged**